source/h1_connection.c

```diff
@@ -704,6 +719,62 @@ static void s_stream_write_outgoing_data(struct h1_stream *stream, struct aws_io
 }
 
 static void s_stream_complete(struct h1_stream *stream, int error_code) {
+    struct h1_connection *connection = AWS_CONTAINER_OF(stream->base.owning_connection, struct h1_connection, base);
+
+    /* Remove stream from list. */
+    aws_linked_list_remove(&stream->node);
+
+    /* If stream completed successfully, check for ways it might alter the state of the connection.
+     * If anything goes wrong here, modify error_code, and the connection will get shut down as a result. */
+    const int original_error_code = error_code;
+    if (!error_code) {
+
+        /* Check whether connection is switching protocols. */
+        if (stream->base.incoming_response_status == AWS_HTTP_STATUS_101_SWITCHING_PROTOCOLS) {
+            /* TODO: confirm that request had sent "Connection: Upgrade" header */
+
+            /* Switching protocols while there are pending streams is too complex to deal with. */
+            bool has_pending_streams = false;
+            if (!aws_linked_list_empty(&connection->thread_data.stream_list)) {
+                has_pending_streams = true;
+            } else {
+                { /* BEGIN CRITICAL SECTION */
+                    int err = aws_mutex_lock(&connection->synced_data.lock);
+                    AWS_FATAL_ASSERT(!err);
+
+                    if (aws_linked_list_empty(&connection->synced_data.pending_stream_list)) {
+                        connection->synced_data.new_stream_error_code = AWS_ERROR_HTTP_SWITCHED_PROTOCOLS;
+                    } else {
+                        has_pending_streams = true;
+                    }
+
+                    err = aws_mutex_unlock(&connection->synced_data.lock);
+                    AWS_FATAL_ASSERT(!err);
+                } /* END CRITICAL SECTION */
+            }
+
+            if (has_pending_streams) {
+                AWS_LOGF_ERROR(
+                    AWS_LS_HTTP_CONNECTION,
+                    "id=%p: Cannot switch protocols while streams are pending, closing connection.",
+                    (void *)&connection->base);
+
+                error_code = AWS_ERROR_INVALID_STATE;
```

Comment:
Why not ```AWS_ERROR_HTTP_SWITCHED_PROTOCOLS```?

Reply:
This is the error for why the connection closed.

The fact that we switched protocols isn't the error, it's that we switched protocols while other HTTP requests were still queued up.

Reply:
That's kinda what I thought, but I had to double check :)